pipeline_dp/pipeline_operations.py

```diff
@@ -6,101 +6,102 @@
-    pass
+    @abc.abstractmethod
+    def values(self, col, stage_name: str):
+        pass
 
-  @abc.abstractmethod
-  def sample_fixed_per_key(self, col, n: int, stage_name: str):
-    pass
+    @abc.abstractmethod
+    def sample_fixed_per_key(self, col, n: int, stage_name: str):
+        pass
 
-  @abc.abstractmethod
-  def count_per_element(self, col, stage_name: str):
-    pass
+    @abc.abstractmethod
+    def count_per_element(self, col, stage_name: str):
+        pass
 
 
 class BeamOperations(PipelineOperations):
-  """Apache Beam adapter."""
+    """Apache Beam adapter."""
 
-  def map(self, col, fn, stage_name: str):
-    return col | stage_name >> beam.Map(fn)
+    def map(self, col, fn, stage_name: str):
+        return col | stage_name >> beam.Map(fn)
 
-  def map_tuple(self, col, fn, stage_name: str):
-    return col | stage_name >> beam.MapTuple(fn)
+    def map_tuple(self, col, fn, stage_name: str):
+        return col | stage_name >> beam.MapTuple(fn)
 
-  def map_values(self, col, fn, stage_name: str):
-    return col | stage_name >> beam.MapTuple(lambda k, v: (k, fn(v)))
+    def map_values(self, col, fn, stage_name: str):
+        return col | stage_name >> beam.MapTuple(lambda k, v: (k, fn(v)))
 
-  def group_by_key(self, col, stage_name: str):
-    return col | stage_name >> beam.GroupByKey()
+    def group_by_key(self, col, stage_name: str):
+        return col | stage_name >> beam.GroupByKey()
 
-  def filter(self, col, fn, stage_name: str):
-    return col | stage_name >> beam.Filter(fn)
+    def filter(self, col, fn, stage_name: str):
+        return col | stage_name >> beam.Filter(fn)
 
-  def keys(self, col, stage_name: str):
-    return col | stage_name >> beam.Keys()
+    def keys(self, col, stage_name: str):
+        return col | stage_name >> beam.Keys()
 
-  def values(self, col, stage_name: str):
-    return col | stage_name >> beam.Values()
+    def values(self, col, stage_name: str):
+        return col | stage_name >> beam.Values()
 
-  def sample_fixed_per_key(self, col, n: int, stage_name: str):
-    return col | stage_name >> combiners.Sample.FixedSizePerKey(n)
+    def sample_fixed_per_key(self, col, n: int, stage_name: str):
+        return col | stage_name >> combiners.Sample.FixedSizePerKey(n)
+
+    def count_per_element(self, col, stage_name: str):
+        return col | stage_name >> combiners.Count.PerElement()
 
-  def count_per_element(self, col, stage_name: str):
-    return col | stage_name >> combiners.Count.PerElement()
 
 class LocalPipelineOperations(PipelineOperations):
-  """Local Pipeline adapter."""
-  
-  def map(self, col, fn, stage_name: str = None):
-    return map(fn, col)
+    """Local Pipeline adapter."""
+
+    def map(self, col, fn, stage_name: str = None):
+        return map(fn, col)
 
-  def map_tuple(self, col, fn, stage_name: str):
-    pass
+    def map_tuple(self, col, fn, stage_name: str = None):
+        return [fn(k, v) for k, v in col]
```

Comment:
We need to have all operations lazy, that's a cool thing that Python often helps with that :). Please use generator expressions https://www.python.org/dev/peps/pep-0289/ (basically replace [...] -> (...) ]

An explanation why async is convenient for us:
1.Having all operations to be async allows to build a pipeline and then to execute it
2.After the pipeline built we can optimize the privacy budget usage (compute the right amount of noise to add) and then to use this noise when the pipeline runs
3.All pipeline frameworks (Beam/Spark) are inherently async - a pipeline is built on the main machine and then executed on workers - which is convenient for us.

Reply:
I forgot about that requirement, thanks for reminding me! 

Should be fixed now